retworkx-core/src/shortest_path/bellman_ford.rs

```diff
@@ -0,0 +1,392 @@
+/// the value is a Vec of node indices of the path starting with `start` and
+/// ending at the index.
+///
+/// Returns a [`DistanceMap`] that maps `NodeId` to path cost if there are no
+/// negative cycles. If there are negative cycles, the result is [`None`].
+/// # Example
+/// ```rust
+/// use retworkx_core::petgraph::Graph;
+/// use retworkx_core::petgraph::prelude::*;
+/// use retworkx_core::dictmap::DictMap;
+/// use retworkx_core::shortest_path::bellman_ford;
+/// use retworkx_core::Result;
+///
+/// let mut graph : Graph<(),i32,Directed>= Graph::new();
+/// let a = graph.add_node(()); // node with no weight
+/// let b = graph.add_node(());
+/// let c = graph.add_node(());
+/// let d = graph.add_node(());
+/// let e = graph.add_node(());
+/// let f = graph.add_node(());
+/// let g = graph.add_node(());
+/// let h = graph.add_node(());
+/// // z will be in another connected component
+/// let z = graph.add_node(());
+///
+/// graph.extend_with_edges(&[
+///     (a, b, 1),
+///     (b, c, 1),
+///     (c, d, 1),
+///     (d, a, 1),
+///     (e, f, -2),
+///     (b, e, -1),
+///     (f, g, 5),
+///     (g, h, 2),
+///     (h, e, -1)
+/// ]);
+/// // a ----> b ----> e ----> f
+/// // ^       |       ^       |
+/// // |       v       |       v
+/// // d <---- c       h <---- g
+///
+/// let expected_res: DictMap<NodeIndex, i32> = [
+///      (a, 3),
+///      (b, 0),
+///      (c, 1),
+///      (d, 2),
+///      (e, -1),
+///      (f, -3),
+///      (g, 2),
+///      (h, 4)
+///     ].iter().cloned().collect();
+/// let res: Result<Option<DictMap<NodeIndex, i32>>> = bellman_ford(
+///     &graph, b, |x| Ok(*x.weight()), None
+/// );
+/// assert_eq!(res.unwrap().unwrap(), expected_res);
+/// // z is not inside res because there is not path from b to z.
+/// ```
+pub fn bellman_ford<G, F, K, E, S>(
+    graph: G,
+    start: G::NodeId,
+    edge_cost: F,
+    mut path: Option<&mut DictMap<G::NodeId, Vec<G::NodeId>>>,
+) -> Result<Option<S>, E>
+where
+    G: IntoEdges + Visitable + NodeIndexable + NodeCount + IntoNodeIdentifiers,
+    G::NodeId: Eq + Hash + IndexType,
+    F: FnMut(G::EdgeRef) -> Result<K, E>,
+    K: Measure + Copy,
+    S: DistanceMap<G::NodeId, K>,
+{
+    let res: BellmanFordData<G, S> = inner_bellman_ford(graph, vec![start], edge_cost)?;
+
+    let BellmanFordData {
+        negative_cycle,
+        scores,
+        predecessor,
+    } = res;
+
+    // Shortest path is not defined
+    if negative_cycle {
+        return Ok(None);
+    }
+
+    // Build path from predecessors
+    if path.is_some() {
+        for node in graph.node_identifiers() {
+            if scores.get_item(node).is_some() {
+                let mut node_path = Vec::<G::NodeId>::new();
+                let mut current_node = node;
+                node_path.push(current_node);
+                while predecessor[current_node.index()].is_some() {
+                    current_node = predecessor[current_node.index()].unwrap();
```

Comment:
We can write this in one line:
```suggestion
                while let Some(current_node) = predecessor[current_node.index()] {
```

Reply:
I will improve it to use the `let Some` syntax, but `let Some(current_node)` shadows `current_node` in the scope hence it causes an infinite loop